OpenNefia.Content/UI/Element/FeatWindow.cs

```diff
@@ -0,0 +1,311 @@
+                    case FeatData.Feat feat:
+                        GraphicsEx.SetColor(Core.Maths.Color.White);
+                        FeatIcons.DrawRegion($"{(int)feat.Prototype.FeatType}", X - 27, Y - 5);
+                        base.Draw();
+                        DescriptionText.Draw();
+                        break;
+                }
+            }
+        }
+
+        protected readonly IPrototypeManager _prototypeManager = default!;
+        private const int ItemsPerPage = 15;
+        private const int WindowHeight = 430;
+        private const int WindowWidth = 740;
+        public UiPagedList<FeatData> List;
+        private UiTextTopic NameTopic;
+        private UiTextTopic DetailTopic;
+        private UiText FeatCountText;
+
+        protected IAssetDrawable AssetInventoryIcons;
+        private IAssetDrawable AssetDecoFeatA;
+        private IAssetDrawable AssetDecoFeatB;
+        private IAssetDrawable AssetDecoFeatC;
+        private IAssetDrawable AssetDecoFeatD;
+
+        private Func<Dictionary<FeatPrototype, int>> GetFeatsFunc;
+        private Action<FeatData.Feat> SelectFeatAction;
+        private Func<int> GetAvailableCountFunc;
+
+        public FeatWindow(Func<Dictionary<FeatPrototype, int>> getFeatsFunc, Action<FeatData.Feat> selectFeatAction, Func<int> getAvailableCountFunc) 
+            : base(keyHintXOffset: 64)
+        {
+            _prototypeManager = IoCManager.Resolve<IPrototypeManager>();
+            AssetInventoryIcons = InventoryHelpers.MakeIcon(InventoryIcon.Feat);
+            AssetDecoFeatA = new AssetDrawable(Protos.Asset.DecoFeatA);
+            AssetDecoFeatB = new AssetDrawable(Protos.Asset.DecoFeatB);
+            AssetDecoFeatC = new AssetDrawable(Protos.Asset.DecoFeatC);
+            AssetDecoFeatD = new AssetDrawable(Protos.Asset.DecoFeatD);
+
+            NameTopic = new UiTextTopic(Loc.GetString("Elona.FeatMenu.NameTopic"));
+            DetailTopic = new UiTextTopic(Loc.GetString("Elona.FeatMenu.DetailTopic"));
+            FeatCountText = new UiText(UiFonts.WindowPage);
+
+            TextTitle.Text = Loc.GetString("Elona.FeatMenu.Title");
+            List = new UiPagedList<FeatData>(ItemsPerPage, this, new Vector2i(-55, -3));
+
+            GetFeatsFunc = getFeatsFunc ?? (() => new Dictionary<FeatPrototype, int>());
+            SelectFeatAction = selectFeatAction ?? (feat => { });
+            GetAvailableCountFunc = getAvailableCountFunc ?? (() => 0);
+
+            List.GrabFocus();
+            List.EventOnActivate += List_OnActivate;
+            RefreshData();
+        }
+
+        private void RefreshData(FeatPrototype? lastSelected = null)
+        {
+            List.Clear();
+            
+            var data = new List<FeatCell>();
+            var prototypes = _prototypeManager.EnumeratePrototypes<FeatPrototype>()?.Where(x => x.FeatType == FeatType.Feat)!;
+            var featCount = GetAvailableCountFunc();
+            var gainedFeats = GetFeatsFunc();
```

Comment:
Rename `GetFeatsFunc()` to `GetGained/Owned/LearnedFeatsFunc()` to make this clearer.

Reply:
changed in b1407d8